streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java

```diff
@@ -252,6 +252,13 @@ private boolean setState(final State newState) {
                 // when the state is already in NOT_RUNNING, its transition to PENDING_SHUTDOWN or NOT_RUNNING (due to consecutive close calls)
                 // will be refused but we do not throw exception here, to allow idempotent close calls
                 return false;
+            } else if (state == State.REBALANCING && newState == State.REBALANCING) {
```

Comment:
I cannot follow here. Why would this happen? Similar for RUNNING?

Reply:
Re first two comments: good call, lgtm.

Re last comment: REBALANCING -> REBALANCING is quite normal, note that before this PR we check this before calling `setState(State.REBALANCING);` so this is prevented, but part of this fix is to move the logic that needs to access the state into a single place (here). Similarly RUNNING -> RUNNING is possible during starting up phase, where we first set the instance state to RUNNING directly to avoid it transit from CREATED -> REBALANCING, and then when threads are starting, it is possible that `maybeSetRunning` went through and hence calls `setState(RUNNING)` again.